pkg/util/pki/parse_test.go

```diff
@@ -174,3 +180,144 @@ func TestDecodePrivateKeyBytes(t *testing.T) {
 		t.Run(test.name, testFn(test))
 	}
 }
+
+type testBundle struct {
+	cert *x509.Certificate
+	pem  []byte
+	pk   crypto.PrivateKey
+}
+
+func mustCreateBundle(t *testing.T, issuer *testBundle, name string) *testBundle {
+	pk, err := GenerateECPrivateKey(256)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	template := &x509.Certificate{
+		Version:               3,
+		BasicConstraintsValid: true,
+		SerialNumber:          serialNumber,
+		PublicKeyAlgorithm:    x509.ECDSA,
+		PublicKey:             pk.Public(),
+		IsCA:                  true,
+		Subject: pkix.Name{
+			CommonName: name,
+		},
+		NotBefore: time.Now(),
+		NotAfter:  time.Now().Add(time.Minute),
+		KeyUsage:  x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+	}
+
+	var (
+		issuerKey  crypto.PrivateKey
+		issuerCert *x509.Certificate
+	)
+
+	if issuer == nil {
+		// Selfsigned (no issuer)
+		issuerKey = pk
+		issuerCert = template
+	} else {
+		issuerKey = issuer.pk
+		issuerCert = issuer.cert
+	}
+
+	certpem, cert, err := SignCertificate(template, issuerCert, pk.Public(), issuerKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return &testBundle{pem: certpem, cert: cert, pk: pk}
+}
+
+func TestParseCertificateChain(t *testing.T) {
+	root := mustCreateBundle(t, nil, "root")
+	int1 := mustCreateBundle(t, root, "int-1")
+	int2 := mustCreateBundle(t, int1, "int-2")
+	leaf := mustCreateBundle(t, int2, "leaf")
+	random := mustCreateBundle(t, nil, "random")
```

Comment:
I may have misunderstood- I thought that this scenario would error- if the input was `{A, B, C, E, G}` from the diagram, we would end up with a list of chain nodes `{A, B, E}`  and `{C, G}` that cannot be merged and would error [here](https://github.com/JoshVanL/cert-manager/blob/parse-certificate-chain-ca/pkg/util/pki/parse.go#L238)?

Reply:
(caveat: I didn't trace through the algorithm in depth with `{A, B, C, E, G}`, because ultimately it'll be unit tests that check this stuff and not humans :grin: )

I guess what I'm saying is that `A->C->G` is a valid chain, and the tree as presented in that diagram is a totally valid PKI hierarchy/tree which could be passed to the function, but we don't have a test which checks its operation.

It's reasonable IMO to do any of the following: 

1. if we can construct two (partial or complete) chains to the same root from given input, we'll error (so `{A, B, C, E, G}` fails because there's `A->C->G` and `A->B->E`)
2. or if we can construct two full chains from root to leaf we'll error (so `{A, B, C, E, G, H}` fails because there's `A->B->E->H` and `A->C->G` but `{A, B, C, E, G}` succeeds with `A->C->G` since `A->B->E` doesn't end in a leaf)
3. or we return all chains we can construct (so `{A, B, C, E, G}` returns `{{A->C->G}, {A->B->E},{A->B},{A->C}}`
4. or we return all complete chains we can construct (so `{A, B, C, E, G}` returns `{{A->C->G}}` and `{A, B, C, E, G, H}` returns `{{A->B->E->H}, {A->C->G}}`)

I think I'd prefer 2 as an implementation, but this is quite academic because in most cases we'd only ever be passed, say `{A, A, B, B, E, H}` which the algorithm should handle fine.

Whichever approach we take, I think my main point is that it would be super valuable to have a test case which takes an input of `{A, B, C, E, G}` and one which takes input of `{A, B, C, E, G, H}`; if they both end up erroring that's fine.

Reply:
Thanks for detailed explanation. Indeed having multiple branches in the same bundle is totally valid and is likely wanted for services with multiple intermediates; they may not even share the same roots, but do trust them. The intention here though is that we expect a single chain when a certificate is _signed_, i.e. the issuer returns the single chain from root to the new signed certificate.

I think managing these larger bundle trees is a separate concern, and falls under the "trust distribution" planned work (whatever that may look like). Given this, I would like to continue with 1 as the implementation. In future it may be that other functions are made to cover those other cases.

To make it clear this is the intention of this function and wouldn't be appropriate for other cases, I'm going to rename it to `ParseSingleCertificateChain`. Open to suggestions for better names :smile:

Reply:
I'll also add some tests to demonstrate that this function should fail with branching chains.